File: python/source/plate.py
import cv2
import numpy as np


class Plate:
    outputSize = (320, 64)

    def __init__(self, frame, rect, roi=None, confidence=0, id=-1):
        self.rect = rect
        self.roi = roi
        self.confidence = confidence
        self.text = None
        self.not_found = 0
        self.id = id
        self.recognized = False

        if roi is not None:
            self.image = self.perspectiveTransform(frame, self.roi)

    def updateImage(self, frame):
        self.image = self.perspectiveTransform(frame, self.roi)

    def perspectiveTransform(self, frame, roi):
        vertices = np.asarray(roi)
        # outputSize = (100, 32)
        outputSize = (320, 64)
        targetVertices = np.array([[0, 0],
                                   [outputSize[0] - 1, 0],
                                   [outputSize[0] - 1, outputSize[1] - 1],
                                   [0, outputSize[1] - 1]
                                   ],
                                  dtype="float32")

        rotationMatrix = cv2.getPerspectiveTransform(vertices, targetVertices)
        result = cv2.warpPerspective(frame, rotationMatrix, outputSize)

        return result

File: cpp/source/plate.cpp
#include <iostream>

#include "plate.h"

using namespace std;


cv::Size Plate::outputSize = cv::Size(320, 64);
cv::Point2f Plate::dst_corners[4] = {cv::Point2f(0, outputSize.height - 1),
                                     cv::Point2f(0, 0),
                                     cv::Point2f(outputSize.width - 1, 0),
                                     cv::Point2f(outputSize.width - 1, outputSize.height - 1)};

Plate::Plate(cv::Mat& image, cv::RotatedRect rotated_rect, vector<cv::Point2f> roi, int id = 0)
{
    this->rotated_rect = rotated_rect;
    this->rect = rotated_rect.boundingRect();
    this->roi = roi;
    this->id = id;
    this->not_found = 0;
    this->recognized = false;

    this->perspectiveTransform(image);
}

Plate::~Plate()
{
    // delete ui;
}

void Plate::perspectiveTransform(cv::Mat& frame)
{
    cv::Point2f* a = &roi[0];
    cv::Mat rotationMatrix = cv::getPerspectiveTransform(a, dst_corners);
    cv::warpPerspective(frame, this->image, rotationMatrix, outputSize);
}
